examples/stabilizer_code.py

```diff
@@ -0,0 +1,192 @@
+
+                # Do the elimination.
+                for k in range(min_row, i):
+                    if M[k, j] == 1:
+                        M[k, :] = np.mod(M[i, :] + M[k, :], 2)
+
+            return rank
+
+        r = _GaussianElimination(M, 0, n - k, 0, n)
+        _ = _GaussianElimination(M, r, n - k, n + r, 2 * n)
+
+        # Get matrix sub-components, as per equation 4.3:
+        # A1 = M[0:r, r : (n - k)]
+        A2 = M[0:r, (n - k) : n]
+        # B = M[0:r, n : (n + r)]
+        C1 = M[0:r, (n + r) : (2 * n - k)]
+        C2 = M[0:r, (2 * n - k) : (2 * n)]
+        # D = M[r : (2 * r), n : (n + r)]
+        E = M[r : (2 * r), (2 * n - k) : (2 * n)]
+
+        X = np.concatenate(
+            [
+                np.zeros((k, r), dtype=np.int8),
+                E.T,
+                np.eye(k, dtype=np.int8),
+                np.mod(E.T @ C1.T + C2.T, 2),
+                np.zeros((k, n - k - r), np.int8),
+                np.zeros((k, k), np.int8),
+            ],
+            axis=1,
+        )
+
+        Z = np.concatenate(
+            [
+                np.zeros((k, n), dtype=np.int8),
+                A2.T,
+                np.zeros((k, n - k - r), dtype=np.int8),
+                np.eye(k, dtype=np.int8),
+            ],
+            axis=1,
+        )
+
+        self.n = n
+        self.k = k
+        self.r = r
+
+        def _BuildByCode(mat):
+            out = []
+            n = mat.shape[1] // 2
+            for i in range(mat.shape[0]):
+                ps = ''
+                for j in range(n):
+                    if mat[i, j] == 0 and mat[i, j + n] == 0:
+                        ps += 'I'
+                    elif mat[i, j] == 1 and mat[i, j + n] == 0:
+                        ps += 'X'
+                    elif mat[i, j] == 0 and mat[i, j + n] == 1:
+                        ps += 'Z'
+                    else:
+                        ps += 'Y'
+                out.append(ps)
+            return out
+
+        self.M = _BuildByCode(M)
+        self.X = _BuildByCode(X)
+        self.Z = _BuildByCode(Z)
+        self.syndromes_to_corrections = {}
+
+        for qid in range(self.n):
+            for op in ['X', 'Z']:
+                syndrome = [
+                    1 if self.M[r][qid] == 'I' or self.M[r][qid] == op else -1
+                    for r in range(self.r)
+                ]
+                self.syndromes_to_corrections[tuple(syndrome)] = (op, qid)
+
+    def encode(self, circuit, qubits):
+        # Equation 4.8:
+        for r, x in enumerate(self.X):
+            for j in range(self.r, self.n - self.k):
+                if x[j] == 'X' or x[j] == 'Y':  # coverage: ignore
+                    circuit.append(  # coverage: ignore
+                        cirq.ControlledOperation(  # coverage: ignore
+                            [qubits[self.n - self.k + r]], cirq.X(qubits[j])  # coverage: ignore
+                        )  # coverage: ignore
+                    )  # coverage: ignore
```

Comment:
This seems like an important distinct path. Could you add test coverage for it?

Reply:
It was indeed. I added a test and it revealed several issues when I was conflating self.r with self.n - self.k

Thanks for the careful review! I am not super confident about my code (my goal was to learn, so I am far from being an expert!).